cmd/kubeadm/app/util/etcd/etcd_test.go

```diff
@@ -494,3 +494,280 @@ func TestClient_GetMemberID(t *testing.T) {
 		})
 	}
 }
+
+func TestListMembers(t *testing.T) {
+	type fields struct {
+		Endpoints      []string
+		newEtcdClient  func(endpoints []string) (etcdClient, error)
+		listMemberFunc func() (*clientv3.MemberListResponse, error)
+	}
+	tests := []struct {
+		name      string
+		fields    fields
+		want      []Member
+		wantError bool
+	}{
+		{
+			name: "PeerURLs are empty",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{}
+					return f, nil
+				},
+			},
+			want: []Member{},
+		},
+		{
+			name: "PeerURLs are non-empty",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								ID:   1,
+								Name: "member1",
+								PeerURLs: []string{
+									"https://member1:2380",
+								},
+							},
+							{
+								ID:   2,
+								Name: "member2",
+								PeerURLs: []string{
+									"https://member2:2380",
+								},
+							},
+						},
+					}
+					return f, nil
+				},
+			},
+			want: []Member{
+				{
+					Name:    "member1",
+					PeerURL: "https://member1:2380",
+				},
+				{
+					Name:    "member2",
+					PeerURL: "https://member2:2380",
+				},
+			},
+		},
+		{
+			name: "PeerURLs has multiple urls",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								ID:   1,
+								Name: "member1",
+								PeerURLs: []string{
+									"https://member1:2380",
+									"https://member2:2380",
+								},
+							},
+						},
+					}
+					return f, nil
+				},
+			},
+			want: []Member{
+				{
+					Name:    "member1",
+					PeerURL: "https://member1:2380",
+				},
+			},
+		},
+		{
+			name: "ListMembers return error",
+			fields: fields{
+				Endpoints: []string{},
+				newEtcdClient: func(endpoints []string) (etcdClient, error) {
+					f := &fakeEtcdClient{
+						members: []*pb.Member{
+							{
+								ID:   1,
+								Name: "member1",
+								PeerURLs: []string{
+									"https://member1:2380",
+									"https://member2:2380",
+								},
+							},
+						},
+					}
+					return f, nil
+				},
+				listMemberFunc: func() (*clientv3.MemberListResponse, error) {
+					return nil, errNotImplemented
+				},
+			},
+			want:      nil,
+			wantError: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := &Client{
+				Endpoints:      tt.fields.Endpoints,
+				newEtcdClient:  tt.fields.newEtcdClient,
+				listMemberFunc: tt.fields.listMemberFunc,
```

Comment:
there is still a problem here:
https://github.com/my-git9/kubernetes/blob/1babd6f0e3fa3f1524e0c938af4c65a832b86993/cmd/kubeadm/app/util/etcd/etcd.go#L275-L301

ideally it should be client.etcdBackoff instead of etcdBackoff.
this will allow us during unit test to pass a very short, one step backoff on &client{....} creation.

you can don't have to do this change for now though, i think. i leave it to your decision.
i'm just trying to explain how to not have tests that create multiple API calls and don't have retries or backoffs.

Reply:
Thanks for your review, can I open a new PR to make this change?

Reply:
i think it's a small change that can be done here in this PR. you can use a second commit if you want.